2_idioms/src/main.rs

```diff
@@ -1,3 +1,452 @@
+            store.two_count += sub;
+            sub = std::cmp::min(left.one_count, right.one_count);
+            left.one_count -= sub;
+            right.one_count -= sub;
+            store.one_count += sub;
+        }
+
+        fn split_money(value: u64, lock_level: u8) -> Money {
+            let mut rest = value;
+            let fifty_count;
+            let twenty_count;
+            let ten_count;
+            let five_count;
+            let two_count;
+            if lock_level <= 1 {
+                fifty_count = rest / 50;
+                rest = rest % 50;
+            } else {
+                fifty_count = 0;
+            }
+            if lock_level <= 2 {
+                twenty_count = rest / 20;
+                rest = rest % 20;
+            } else {
+                twenty_count = 0;
+            }
+            if lock_level <= 3 {
+                ten_count = rest / 10;
+                rest = rest % 10;
+            } else {
+                ten_count = 0;
+            }
+            if lock_level <= 4 {
+                five_count = rest / 5;
+                rest = rest % 5;
+            } else {
+                five_count = 0;
+            }
+            if lock_level < 5 {
+                two_count = rest / 2;
+                rest = rest % 2;
+            } else {
+                two_count = 0;
+            }
+            let one_count = rest;
+            Money {
+                fifty_count,
+                twenty_count,
+                ten_count,
+                five_count,
+                two_count,
+                one_count,
+            }
+        }
+    }
+}
+
+mod product {
+    use crate::trade;
+    use std::any::Any;
+    pub trait Product: trade::HasValue + Any {
+        fn name(&self) -> &'static str;
+    }
+    #[derive(Hash, Debug)]
+    //Amount of items remaining
+    pub struct Amount(pub u32);
+    impl Into<Amount> for u32 {
+        fn into(self) -> Amount {
+            Amount(self)
+        }
+    }
+}
+struct Stock {
+    stock_capacity: RowCount,
+    in_stock_capacity: InRowCount,
+    product_storage: std::collections::HashMap<TypeId, Vec<Box<dyn Product>>>,
+}
+
+impl Stock {
+    pub fn new<T1: Into<RowCount>, T2: Into<InRowCount>>(
+        stock_capacity: T1,
+        in_stock_capacity: T2,
+    ) -> Self {
+        Stock {
+            stock_capacity: stock_capacity.into(),
+            in_stock_capacity: in_stock_capacity.into(),
+            product_storage: std::collections::HashMap::new(),
+        }
+    }
+
+    pub fn display_stock(&self) {
+        self.product_storage.iter().for_each(|kvp| {
+            let product = kvp.1;
+            println!(
+                "product:{}, price:{}, in stock:{:?};",
+                product[0].name(),
+                product[0].value(),
+                product.len()
+            )
+        })
+    }
+
+    pub fn insert<T: Product>(&mut self, items: Vec<Box<T>>) -> Result<(), &'static str> {
```

Comment:
@ArsileLuci stringly-typed errors are not idiomatic, nor are convenient. Use enums for that and have strongly-typed errors.